src/procedures/utility/build_weighted_matrix.c

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright FalkorDB Ltd. 2023 - present
+ * Licensed under the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "./internal.h"
+
+// the ith relationID is i if no relation is given, and rels[i] if it is.
+#define GETRELATIONID(i) ((rels) ? rels[i] : i)
+
+// frees the build weighted matrix workspace
+#define BWM_FREE                                                 \
+{                                                                \
+	if (weight != ATTRIBUTE_ID_NONE) GrB_BinaryOp_free(&minID);  \
+	GrB_Scalar_free(&theta);                                     \
+	GrB_Type_free(&contx_type);                                  \
+	GrB_UnaryOp_free(&toMatrix);                                 \
+	GrB_BinaryOp_free(&weightOp);                                \
+	GrB_BinaryOp_free(&toMatrixMin);                             \
+	GxB_IndexBinaryOp_free(&minID_indexOP);                      \
+}
+
+// structure that holds all the context nessesary for the GraphBLAS functions
+// can select the right edge
+typedef struct
+{
+	const Graph *g;  // Graph
+	AttributeID w;   // Attribute used as weight
+	int comp;        // -1 if max, 1 if min
+} compareContext;
+
+// binary index op, does not use the index but needs the context. 
+// will compare two edges and give the one with the minimum or maximum weight.
+static void _compare_EdgeID_value 
+(
+	uint64_t *z,
+	const uint64_t *x,
+	GrB_Index ix,
+	GrB_Index jx,
+	const uint64_t *y,
+	GrB_Index iy,
+	GrB_Index jy,
+	const compareContext *ctx
+) {
+	*z = *x;
+	Edge _x, _y;
+	Graph_GetEdge(ctx->g, (EdgeID) (*x), &_x);
+	Graph_GetEdge(ctx->g, (EdgeID) (*y), &_y);
+	SIValue *xv = GraphEntity_GetProperty((GraphEntity *) &_x, ctx->w);
+	SIValue *yv = GraphEntity_GetProperty((GraphEntity *) &_y, ctx->w);
+
+	if((SI_TYPE(*xv) & SI_NUMERIC) == 0 || 
+		((SI_TYPE(*yv) & SI_NUMERIC) && 
+		SIValue_Compare(*xv, *yv, NULL) == ctx->comp)
+	) {
+		*z = *y;
+	}
```

Comment:
This part need explanations,

Either `xv` is not numeric in which case the output would be `y` but what if `y` isn't numeric ?
Or `yv` is numeric in that case we call `SIValue_Compare` and if `yv` is either greater or smaller (depending on the context) we set `z` to `y`.